backend/src/Edna/Dashboard/Service.hs

```diff
@@ -0,0 +1,130 @@
+  , deleteSubExperiment
+  , getExperiments
+  , getSubExperiment
+  , getMeasurements
+  ) where
+
+import Universum
+
+import qualified Data.HashSet as HS
+
+import Database.Beam.Backend.SQL.BeamExtensions (unSerial)
+import Database.Beam.Postgres (PgJSON(..))
+import Fmt (fmt, (+|), (|+))
+import Servant.API (NoContent(..))
+
+import qualified Edna.Dashboard.DB.Query as Q
+
+import Edna.Analysis.FourPL (Params4PL(..))
+import Edna.Dashboard.DB.Schema (MeasurementT(..), SubExperimentT(..))
+import Edna.Dashboard.Error (DashboardError(..))
+import Edna.Dashboard.Web.Types
+  (ExperimentResp(..), ExperimentsResp(..), MeasurementResp(..), SubExperimentResp(..))
+import Edna.Logging (logMessage)
+import Edna.Setup (Edna)
+import Edna.Util
+  (CompoundId, IdType(..), ProjectId, SubExperimentId, TargetId, justOrThrow, unSqlId)
+import Edna.Web.Types (WithId(..))
+
+-- | Make given sub-experiment the primary one for its parent experiment.
+-- Previous primary sub-experiment is no longer primary.
+makePrimarySubExperiment :: SubExperimentId ->
+  Edna (WithId 'SubExperimentId SubExperimentResp)
+makePrimarySubExperiment subExpId = do
+  logMessage $ fmt $ "Making sub-experiment " +| subExpId |+ " primary"
+  Q.makePrimarySubExperiment subExpId
+  getSubExperiment subExpId
+
+-- | Update name of a sub-experiment.
+setNameSubExperiment :: SubExperimentId -> Text ->
+  Edna (WithId 'SubExperimentId SubExperimentResp)
+setNameSubExperiment subExpId name = do
+  logMessage $ fmt $ "Seetting the name of sub-experiment " +| subExpId |+
+    " to " +| name |+ ""
+  Q.setNameSubExperiment subExpId name
+  getSubExperiment subExpId
+
+-- | Update @isSuspicious@ flag for a sub-experiment.
+setIsSuspiciousSubExperiment :: SubExperimentId -> Bool ->
+  Edna (WithId 'SubExperimentId SubExperimentResp)
+setIsSuspiciousSubExperiment subExpId isSuspicious = do
+  logMessage $ fmt $ "Marking sub-experiment " +| subExpId |+ mappend " as "
+    if isSuspicious then "suspicious" else "not suspicious"
+  Q.setIsSuspiciousSubExperiment subExpId isSuspicious
+  getSubExperiment subExpId
+
+-- | Delete a sub-experiment with given ID. Fails if this sub-experiment is
+-- primary.
+deleteSubExperiment :: SubExperimentId -> Edna NoContent
+deleteSubExperiment subExpId = NoContent <$ do
+  logMessage $ fmt $ "Deleting sub-experiment " +| subExpId |+ ""
+  whenM (Q.isPrimarySubExperiment subExpId) $
+    throwM $ DECantDeletePrimary subExpId
+
+  Q.deleteSubExperiment subExpId
+
+-- | Get data about all experiments using 3 optional filters: by project ID,
+-- compound ID and target ID. If filters by compound and target are specified,
+-- also compute average IC50 for them.
+getExperiments :: Maybe ProjectId -> Maybe CompoundId -> Maybe TargetId ->
+  Edna ExperimentsResp
+getExperiments mProj mComp mTarget = do
+  pairs <- Q.getExperiments mProj mComp mTarget
+  meanIC50 <- case (mComp, mTarget) of
+    (Just _, Just _) -> Just <$> computeMeanIC50 (map snd pairs)
```

Comment:
What mean `MeanIC50` for scientists?

Reply:
AFAIU the person who decides whether a compound is good or not is interested in IC50 value for `(compound, target that is supposed to be attacked by this compound)`. If they select this compound and target in dashboard, they will see potentially many experiments with this pair and values may be different in these experiments (AFAIU in reality values should be similar, but they can't be absolutely equal). But it's hard to make a decision when you see 10 values, so we also provide average value.

I might be wrong, but I vaguely remember "mean IC50" being mentioned on some meetings, and that's how it appeared in designs.